contracts/sol6/reserves/KyberFprReserveV2.sol

```diff
@@ -0,0 +1,357 @@
+                return token.balanceOf(address(this));
+            }
+            usingToken = token;
+        }
+
+        uint256 balanceOfWallet = usingToken.balanceOf(wallet);
+        uint256 allowanceOfWallet = usingToken.allowance(wallet, address(this));
+
+        return minOf(balanceOfWallet, allowanceOfWallet);
+    }
+
+    /// @dev return wallet that holds the token
+    ///      if token is ETH, check tokenWallet of WETH instead
+    ///      if wallet is 0x0, consider as this reserve address
+    function getTokenWallet(IERC20 token) public view returns (address wallet) {
+        wallet = (token == ETH_TOKEN_ADDRESS)
+            ? tokenWallet[address(weth)]
+            : tokenWallet[address(token)];
+        if (wallet == address(0)) {
+            wallet = address(this);
+        }
+    }
+
+    /// @dev do a trade, re-validate the conversion rate, remove trust assumption with network
+    /// @param srcToken Src token
+    /// @param srcAmount Amount of src token
+    /// @param destToken Destination token
+    /// @param destAddress Destination address to send tokens to
+    function doTrade(
+        IERC20 srcToken,
+        uint256 srcAmount,
+        IERC20 destToken,
+        address payable destAddress,
+        uint256 conversionRate
+    ) internal {
+        require(conversionRate > 0, "rate is 0");
+
+        uint256 rate;
+
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            require(msg.value == srcAmount, "wrong msg value");
+            rate = conversionRatesContract.getRate(destToken, block.number, true, srcAmount);
+        } else {
+            require(msg.value == 0, "bad msg value");
+            rate = conversionRatesContract.getRate(srcToken, block.number, false, srcAmount);
+        }
+
+        // re-validate conversion rate
+        require(rate >= conversionRate, "rate is lower than conversion rate");
+        if (sanityRatesContract != IKyberSanity(0)) {
+            // sanity rate check
+            uint256 sanityRate = sanityRatesContract.getSanityRate(srcToken, destToken);
+            require(rate <= sanityRate, "rate should not be greater than sanity rate" );
+        }
+
+        uint256 destAmount = calcDestAmount(srcToken, destToken, srcAmount, conversionRate);
+        require(destAmount > 0, "dest amount is 0");
+
+        address srcTokenWallet = getTokenWallet(srcToken);
+        address destTokenWallet = getTokenWallet(destToken);
+
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            // add to imbalance
+            conversionRatesContract.recordImbalance(
+                destToken,
+                int256(destAmount),
+                0,
+                block.number
+            );
+            // if reserve is using weth, convert eth to weth and transfer weth to its' tokenWallet
+            if (srcTokenWallet != address(this)) {
+                weth.deposit{value: msg.value}();
```

Comment:
so actually the assumption here is that if its weth its external wallet.
and if its an external for Eth its using WETH

I mean
no option to use weth and have wallet address to be this.

its good. 
just not sure its totally clear.

I think we can add another @dev note in setTokenWallet

Reply:
There is no reason of using weth in this reserve address right? we can not approve other wallets to get weth from reserve

Reply:
I will add a dev note in setTokenWallet

Reply:
yes
no reason.
just to clarify